### DISCUSSION TITLE
Design a new prefetcher for fetching data from main memory to cache hierarchy.

### DISCUSSION THREAD

**Discussion by szuhzw** [CMU-SAFARI/Hermes#30] [Q&A] [answered]
Hello. Which files do I need to modify in the Hermes project if I want to design a new prefetcher for fetching data from main memory to cache hierarchy? Thank you. @rahulbera

**Comment by rahulbera** [comment, marked as answer]
There are two ways to do this:
- Option 1: (a) Define your own prefetcher in a separate file. E.g., if it's a L2 prefetcher, define it as `myprefetcher.l2c_pref`, if it's an LLC prefetcher, call it `myprefetcher.llc_pref`. (b) Build ChampSim with this prefetcher of choice (i.e., pass the prefetcher name in the `build_champsim.sh` script). See [ChampSim documentation](https://champsim.github.io/ChampSim/master/Modules.html#memory-prefetchers) to know more.
- Option 2: Define the internal logic of your prefetcher as we have done in `scooby.cc`. See the `prefetcher/streamer.cc` file for an easiest example. All you have to do is to (a) create a class for your prefetcher as a dervied class of `Prefetcher` (see `inc/streamer.h` for example), (b) define three key functions for this new prefetcher's operation: `invoke_prefetcher`, `print_config`, and `dump_stats`, (c) call this prefetcher from the `multi` prefetcher interface: in case for L2/LLC prefetcher, call it from `multi.l2c_pref` or `multi.llc_pref`, respectively.

The first option binds the prefetcher at compile time. So if you have to try and test different prefetchers, you have to compile different versions of ChampSim with appropriate prefetchers. The second option binds a prefetcher at runtime: you just need to compile once and select which prefetcher to use at runtime using `l2c_prefetcher_types` or `llc_prefetcher_types` knobs. But the option 2, due to it's runtime bind, may run a bit slower than option 1.
